src/server/node_manager.c

```diff
@@ -6839,6 +6839,51 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
 	return 0;
 }
 
+void
+disassociate_job_from_moms(job *pjob, struct pbsnode *pnode)
+{
+	int i;
+	int j;
+	int special_case = 0;
+	mom_svrinfo_t *psvrmom;
+	
+	if (pnode == NULL)
+		return;
+			
+	for (i = 0; i < pnode->nd_nummoms; i++) {	
+		if (pnode->nd_moms[i] == NULL)	
+			continue;	
+		psvrmom = (mom_svrinfo_t *)(pnode->nd_moms[i]->mi_data);	
+
+		for (j=0; j<psvrmom->msr_jbinxsz; j++) {	
```

Comment:
I think we should work on this and try to get rid of this loop also, as it is going to all jobs assigned to mom

Reply:
In general, its good to remove the loop. But here we need to consider this case in detail. 
Whenever a job queued in MoM, the job object would be registered in mominfo structure and the same would be removed after completion. For short jobs, it is so quick.  Even if we try with AVL tree DS to store job id versus mominfo index, the operation would be more complex in adding/removing/ frequently. The gain we get on from that would be trivial. (We are already getting 93% to 0.5 %).  I feel those operations are effective in terms of a large number of elements like tens of hundreds of thousands. 

This in case, at any moment, the number of runtime jobs is close to a few hundred, I would prefer the current code is much better in comparison to having another DS. If you think otherwise, plz explain with a technical explanation, I will consider in another PR.

Reply:
Thanks, @bremanandjk for the explanation, but my comment was in a generic or architectural way (sorry if it was not clear enough), as currently our mom structures are using array and looping through those arrays, which is slow, we can refactor those and make them faster.